python/plugins/processing/algs/qgis/Aggregate.py

```diff
@@ -0,0 +1,223 @@
+        aggregates = self.getParameterValue(self.AGGREGATES)
+        dissolve = self.getParameterValue(self.DISSOLVE)
+        output = self.getOutputFromName(self.OUTPUT)
+
+        layer = dataobjects.getObjectFromUri(layer)
+
+        self.da = QgsDistanceArea()
+        self.da.setSourceCrs(layer.crs())
+        self.da.setEllipsoidalMode(True)
+        self.da.setEllipsoid(QgsProject.instance().readEntry(
+            'Measure', '/Ellipsoid', GEO_NONE)[0])
+        self.exp_context = layer.createExpressionContext()
+
+        group_by_expr = self.createExpression(group_by)
+
+        geometry_expr = self.createExpression('collect($geometry, {})'
+                                              .format(group_by))
+        fields = []
+        fields_expr = []
+        for field_def in aggregates:
+            fields.append(QgsField(name=field_def['name'],
+                                   type=field_def['type'],
+                                   len=field_def['length'],
+                                   prec=field_def['precision']))
+            aggregate = field_def['aggregate']
+            if aggregate == 'first_value':
+                expression = field_def['input']
+            elif aggregate == 'concatenate':
+                expression = ('{}({}, {}, {}, \'{}\')'
+                              .format(field_def['aggregate'],
+                                      field_def['input'],
+                                      group_by,
+                                      'TRUE',
+                                      field_def['delimiter']))
+            else:
+                expression = '{}({}, {})'.format(field_def['aggregate'],
+                                                 field_def['input'],
+                                                 group_by)
+            expr = self.createExpression(expression)
+            fields_expr.append(expr)
+
+        if dissolve:
+            geomType = QgsWkbTypes.singleType(layer.wkbType())
+        else:
+            geomType = QgsWkbTypes.multiType(layer.wkbType())
+        writer = output.getVectorWriter(fields, geomType, layer.crs())
+
+        count_distinct_expr = self.createExpression('count_distinct({})'
+                                                    .format(group_by))
+        total = self.evaluateExpression(count_distinct_expr)
+        progress_step = 100.0 / (total or 1)
+        processed_values = []
+
+        feature = QgsFeature()
+        features = vector.features(layer)
+        for feature in features:
+            self.exp_context.setFeature(feature)
+            # row_number doesn't really make sense here
+            #exp_context.lastScope().setVariable("row_number", current + 1)
+
+            group_by_value = self.evaluateExpression(group_by_expr)
+            if group_by_value in processed_values:
+                continue
+            processed_values.append(group_by_value)
+
+            geometry = self.evaluateExpression(geometry_expr)
```

Comment:
I have concerns with this approach. What you're doing here is:

1. iterating over every feature in the layer
2. for every group value, iterating `(1 * len(aggregates))` over the entire layer again

This means selecting 2 aggregates results in 4 iterations over the entire layer. And this is the ideal case, where the group by clause can be fully compiled and handed off to the provider. If it can't be, then you'll be iterating `1 + number of distinct group by values * (1+len(aggregates))` times over the whole layer (since the entire layer would be iterated to determine whether each feature falls into the current group each time). That's going to hurt...

Alternative approaches would be:
1. fetch all the features upfront, store in a list, and then reuse them (bad on memory)
2. iterate the outer loop once to determine each group value (but tweak the filter request for this so it only includes attributes required for the group by expression + geometry if needed for group by expression). For each unique group by value, refetch matching features from the provider and store in a list. Calculate the collection & aggregates over this list.
3. Maybe changing processing to use a QgsVectorLayerCache and having vector.features return a QgsCachedFeatureIterator/QgsCachedFeatureWriterIterator would help. That could avoid fetching the same features from the provider multiple times.